Neos.ContentRepository.Core/Classes/EventStore/EventsToPublish.php

```diff
@@ -33,4 +35,31 @@ public static function empty(): self
             ExpectedVersion::ANY()
         );
     }
+
+    public function withCausationOfFirstEventAndAdditionalMetaData(EventMetadata $metadata): self
+    {
+        /** @var list<EventInterface|DecoratedEvent> $restEvents */
+        $restEvents = iterator_to_array($this->events);
+        if (empty($restEvents)) {
+            return $this;
+        }
+        $firstEvent = array_shift($restEvents);
+
+        if ($firstEvent instanceof DecoratedEvent && $firstEvent->eventMetadata) {
+            $metadata = EventMetadata::fromArray(array_merge($firstEvent->eventMetadata->value, $metadata->value));
+        }
+
+        $decoratedFirstEvent = DecoratedEvent::create($firstEvent, eventId: Event\EventId::create(), metadata: $metadata);
+
+        $decoratedRestEvents = [];
+        foreach ($restEvents as $event) {
+            $decoratedRestEvents[] = DecoratedEvent::create($event, causationId: $decoratedFirstEvent->eventId);
+        }
+
+        return new EventsToPublish(
+            $this->streamName,
+            Events::fromArray([$decoratedFirstEvent, ...$decoratedRestEvents]),
+            $this->expectedVersion
+        );
```

Comment:
That's quite a lot of complexity. I don't really think that we need to handle the first event differently from the others in this case (see suggestion below) so maybe we can simplify this to a one-liner like:
```php
return new self(
    $this->streamName,
    $this->events->map(fn (Event $event) => DecoratedEvent::create(
        event: $event,
        correlationId: $correlationId,
    ),
    $this->expectedVersion,
);
```